### PR TITLE
DetectorInfo: sync scan support for better performance

### LINKED ISSUE #20027: Add convert to scanning method to DetectorInfo
The proposal here is to add a method to `DetectorInfo` to convert a static `DetectorInfo` into a time indexded one, by passing in a vector of start time, end time pairs. The pairs are checked to ensure they do not overlap, and then the same time ranges are set for every detector.

The rationale behind this is that the `build()` step in `ScanningWorkspaceBuilder` is currently very slow for D2B and D20 at the ILL. An example of this can be seen in the system test added in PR https://github.com/mantidproject/mantid/pull/19920 (in progress). Attached is the profiling output that corresponds to this test. It can be seen that the merging of the `DetectorInfo` objects takes the most time, this ticket would avoid the need to merge the `DetectorInfo` objects.

[cpu_profile_standard.pdf](https://github.com/mantidproject/mantid/files/1155741/cpu_profile_standard.pdf)

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Add methods for setting up `DetectorInfo` for a sync scan of all detectors. This avoid a big overhead when merging many `DetectorInfo` objects into one big scan.

See also https://github.com/mantidproject/mantid/pull/20204 (replaced by this PR).

**To test:**

Code review.

Try loading the attached file for D20 with and without the changes, preferably on a release branch. Check the speed up (on my PC 15.8s without changes, 3s with).

[d20_det_scan.zip](https://github.com/mantidproject/mantid/files/1211628/d20_det_scan.zip)

Fixes #20027.

**To test:**

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by ianbush — COMMENTED** [review]
Generally this looks good, and I see the speedup promised. I will have a double check through before approving once the merge conflicts are fixed.

File: Framework/Geometry/src/Instrument/DetectorInfo.cpp
Comment (ianbush) [inline Framework/Geometry/src/Instrument/DetectorInfo.cpp]: `->`?

Comment (SimonHeybrock) [inline Framework/Geometry/src/Instrument/DetectorInfo.cpp]: Thanks.

**Review by ianbush — APPROVED** [review]
Looks good, just needs the builds to pass then good to merge.

**Pull request merged**
